src/pubkey.ts

```diff
@@ -0,0 +1,209 @@
+}
+
+/**
+ * ECDSA public key recovery from signature.
+ * @returns Recovered public key
+ */
+export const ecrecover = function(
+  msgHash: Buffer,
+  v: number,
+  r: Buffer,
+  s: Buffer,
+  chainId?: number,
+): Buffer {
+  const signature = Buffer.concat([setLength(r, 32), setLength(s, 32)], 64)
+  const recovery = calculateSigRecovery(v, chainId)
+  if (!isValidSigRecovery(recovery)) {
+    throw new Error('Invalid signature v value')
+  }
+  const senderPubKey = secp256k1.recover(msgHash, signature, recovery)
+  return secp256k1.publicKeyConvert(senderPubKey, false).slice(1)
+}
+
+/**
+ * Convert signature parameters into the format of `eth_sign` RPC method.
+ * @returns Signature
+ */
+export const toRpcSig = function(v: number, r: Buffer, s: Buffer, chainId?: number): string {
+  const recovery = calculateSigRecovery(v, chainId)
+  if (!isValidSigRecovery(recovery)) {
+    throw new Error('Invalid signature v value')
+  }
+
+  // geth (and the RPC eth_sign method) uses the 65 byte format used by Bitcoin
+  return bufferToHex(Buffer.concat([setLengthLeft(r, 32), setLengthLeft(s, 32), toBuffer(v)]))
+}
+
+/**
+ * Convert signature format of the `eth_sign` RPC method to signature parameters
+ * NOTE: all because of a bug in geth: https://github.com/ethereum/go-ethereum/issues/2053
+ */
+export const fromRpcSig = function(sig: string): ECDSASignature {
+  const buf: Buffer = toBuffer(sig)
+
+  // NOTE: with potential introduction of chainId this might need to be updated
+  if (buf.length !== 65) {
+    throw new Error('Invalid signature length')
+  }
+
+  let v = buf[64]
+  // support both versions of `eth_sign` responses
+  if (v < 27) {
+    v += 27
+  }
+
+  return {
+    v: v,
+    r: buf.slice(0, 32),
+    s: buf.slice(32, 64),
+  }
+}
+
+/**
+ * Validate a ECDSA signature.
+ * @param homesteadOrLater Indicates whether this is being used on either the homestead hardfork or a later one
+ */
+export const isValidSignature = function(
+  v: number,
+  r: Buffer,
+  s: Buffer,
+  homesteadOrLater: boolean = true,
+  chainId?: number,
+): boolean {
+  const SECP256K1_N_DIV_2 = new BN(
+    '7fffffffffffffffffffffffffffffff5d576e7357a4501ddfe92f46681b20a0',
+    16,
+  )
+  const SECP256K1_N = new BN('fffffffffffffffffffffffffffffffebaaedce6af48a03bbfd25e8cd0364141', 16)
+
+  if (r.length !== 32 || s.length !== 32) {
+    return false
+  }
+
+  if (!isValidSigRecovery(calculateSigRecovery(v, chainId))) {
+    return false
+  }
+
+  const rBN: BN = new BN(r)
+  const sBN: BN = new BN(s)
+
+  if (rBN.isZero() || rBN.gt(SECP256K1_N) || sBN.isZero() || sBN.gt(SECP256K1_N)) {
+    return false
+  }
+
+  if (homesteadOrLater && sBN.cmp(SECP256K1_N_DIV_2) === 1) {
+    return false
+  }
+
+  return true
+}
+
+function calculateSigRecovery(v: number, chainId?: number): number {
+  return chainId ? v - (2 * chainId + 35) : v - 27
+}
+
+function isValidSigRecovery(recovery: number): boolean {
+  return recovery === 0 || recovery === 1
+}
```

Comment:
I would have a tendency to move to top-most functions to ``address.js`` and make for the others some dedicated ``signature.js`` file.